torch/autograd/graph.py

```diff
@@ -130,3 +132,81 @@ def unpack_from_cpu(packed):
             return tensor.to(device, non_blocking=pin_memory)
 
         super().__init__(pack_to_cpu, unpack_from_cpu)
+
+
+_cloned = dict()
+_use_counts = dict()
+_keep_graph = False
+class _swap_with_cloned(saved_tensors_hooks):
+    def __init__(self):
+        def pack_hook(t):
+            uid = (t.data_ptr(), t._version)
+            if not _keep_graph:
+                _use_counts[uid] = _use_counts.get(uid, 0) + 1
+            return uid, t
+
+        def unpack_hook(tup):
+            uid, t = tup
+            if uid in _cloned:
+                res = _cloned[uid]
+            else:
+                res = t
+            if not _keep_graph:
+                if uid not in _use_counts:
+                    raise RuntimeError("If you are trying to backward through the graph a second time "
+                                       "or trying to compute higher-order gradients, please specify "
+                                       "`keep_graph=True` when enabling allow_mutation_on_saved_tensors")
+                _use_counts[uid] -= 1
+                if _use_counts[uid] == 0:
+                    if uid in _cloned:
+                        del _cloned[uid]
+                    del _use_counts[uid]
+            return res
+        super().__init__(pack_hook, unpack_hook)
+
+class _CloneArgBeforeMutateMode(TorchDispatchMode):
+    @staticmethod
+    def is_mutating(func):
+        # We may want to also handle out= later
+        return func.__name__.split('.')[0][-1] == "_"
```

Comment:
I wonder if there is some introspection we can do to figure out (1) whether the op is mutating any arguments, and (2) which arguments are being mutated